packages/flutter/lib/src/material/focusable.dart

```diff
@@ -0,0 +1,297 @@
+  /// set, unless it is the descendant of another scope.
+  final bool autofocus;
+
+  /// True when displaying focus decorations.
+  ///
+  /// Set to false if you wish to handle the visual focus indicators yourself.
+  ///
+  /// If you just want a different decoration around the focusable widget, set
+  /// [focusedDecoration] and/or [unfocusedDecoration], or modify the
+  /// [FocusableThemeData] in the ambient [Theme].
+  ///
+  /// Defaults to true.
+  final bool showDecorations;
+
+  /// Defines the decoration to be drawn around the [child] when this widget
+  /// receives the focus.
+  ///
+  /// This defaults to the [FocusableThemeData.focusedDecoration] provided by
+  /// the ambient [Theme] in the context.
+  final Decoration focusedDecoration;
+
+  /// Defines the decoration to be drawn around the [child] when this widget
+  /// does not have the focus.
+  ///
+  /// This defaults to the [FocusableThemeData.unfocusedDecoration] provided by
+  /// the ambient [Theme] in the context.
+  final Decoration unfocusedDecoration;
+
+  /// Returns the [node] of the [Focusable] that most tightly encloses the given
+  /// [BuildContext].
+  ///
+  /// The [context] argument must not be null.
+  static FocusableNode of(BuildContext context) {
+    assert(context != null);
+    final _FocusableMarker marker = context.inheritFromWidgetOfExactType(_FocusableMarker);
+    return marker?.node ?? context.owner.focusManager.rootFocusable;
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(StringProperty('debugName', debugLabel, defaultValue: null));
+    properties.add(DiagnosticsProperty<Decoration>('decoration', focusedDecoration));
+  }
+
+  @override
+  _FocusableState createState() => _FocusableState();
+}
+
+class _FocusableState extends State<Focusable> {
+  FocusableNode node;
+  bool _hasFocus;
+
+  @override
+  void initState() {
+    super.initState();
+    node = FocusableNode(
+      isScope: widget.isScope,
+      autofocus: widget.autofocus,
+      key: GlobalKey(debugLabel: widget.debugLabel),
+    );
+    _hasFocus = node.hasFocus;
+    node.addListener(_handleFocusChanged);
+  }
+
+  @override
+  void dispose() {
+    node.dispose();
+    super.dispose();
+  }
+
+  void _handleFocusChanged() {
+    if (_hasFocus != node.hasFocus) {
+      setState(() {
+        _hasFocus = node.hasFocus;
+      });
+      if (widget.onFocusChange != null) {
+        widget.onFocusChange(node.hasFocus);
+      }
+    }
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    final FocusableNode newParent = Focusable.of(context);
+    newParent.reparent(node);
+  }
+
+  Decoration _getDecoration(BuildContext context) {
+    if (!widget.showDecorations) {
+      return null;
+    }
+    if (_hasFocus) {
+      return widget.focusedDecoration ?? Theme.of(context).focusableTheme.focusedDecoration;
```

Comment:
assuming we move this to widgets, we'll have to have another solution here.
I think a reasonable one would be for every widget in the material layer to fetch the decorations from the theme themselves and fill it in the Focusable. Another option would be for the material layer to create a MaterialFocusable.

Reply:
another option would be to use a FocusableTheme widget, similar to IconTheme.

Reply:
Or how about separating out a FocusHighlight widget that does the highlighting separately from the Focusable (it is a descendant of the Focusable and listens to the focus changes directly from the node).

Reply:
makes sense. how do you make that work without the author having to own the focus node, though? Do you have an equivalent of ye olde `Focus.at` in this world? (`Focus.at` was an early API that would tell you if the nearest focused scope had focus.)